apps/voting/src/shared/store.ts

```diff
@@ -119,48 +122,115 @@ const electionConfigBaseAtom = atom<ElectionConfig | undefined>(undefined);
+                description: contractConfig.election_description,
+                manifest: contractConfig.election_manifest,
+                parameters: contractConfig.election_parameters,
+                candidates: contractConfig.candidates,
+            };
         }
 
-        const electionManifestPromise = getChecksumResource<ElectionManifest>(config.election_manifest);
-        const electionParametersPromise = getChecksumResource<ElectionParameters>(config.election_parameters);
-        const candidatePromises = config.candidates.map(getCandidate);
-
+        const electionManifestPromise = getChecksumResource<ElectionManifest>(preliminaryConfig.manifest);
+        const electionParametersPromise = getChecksumResource<ElectionParameters>(preliminaryConfig.parameters);
+        const candidatePromises = preliminaryConfig.candidates.map(getCandidate);
         const [manifest, parameters, ...candidates] = await Promise.all([
             electionManifestPromise,
             electionParametersPromise,
             ...candidatePromises,
         ]);
 
-        const mappedConfig: ElectionConfig = {
-            start: Timestamp.toDate(config.election_start),
-            end: Timestamp.toDate(config.election_end),
+        const config: ElectionConfig = {
+            ...preliminaryConfig,
             candidates: candidates.filter(isDefined),
-            description: config.election_description,
-            manifest,
-            parameters,
-            guardianKeys: config.guardian_keys,
-            setupDone: false,
+            manifest: manifest,
+            parameters: parameters,
         };
+        set(electionConfigBaseAtom, config);
+    },
+);
 
-        set(electionConfigBaseAtom, mappedConfig);
+/**
+ * Primitive atom for holding the {@linkcode ElectionConfig} of the election contract
+ */
+const guardiansStateBaseAtom = atom<GuardiansState | undefined>(undefined);
 
-        const guardians = await getGuardiansState();
-        if (guardians === undefined) {
-            throw new Error('Failed to get election config');
+/**
+ * Holds the configuration of the election contract. A reference to this should always be kept in the application root
+ * to avoid having to fetch the configuration more than once.
+ */
+export const guardiansStateAtom = atom(
+    (get) => get(guardiansStateBaseAtom),
+    async (_, set) => {
+        const hydrate = CONFIG.contractConfig !== undefined;
+        let config: GuardiansState;
+        if (hydrate) {
+            config = {
+                guardianKeys: CONFIG.contractConfig!.guardian_keys,
+                setupDone: CONFIG.contractConfig!.guardians_setup_done,
+            };
+        } else {
+            const guardians = await getGuardiansState();
+            if (guardians === undefined) {
+                throw new Error('Failed to get election config');
+            }
+
+            const setupDone = guardians.every(
```

Comment:
Just for my understanding of the timeline/process. The smart contracts should be set up and the guardians should register themself (hopefully successfully) and only after that we start e.g. our front end. Otherwise (if the guardian state is only fetched once), we could end up that the guardian state in the front end shows `setupDone==false` (and people have to manually reload the page to fetch/refresh it again).

Reply:
The guardians state is only fetched (both by backend and frontend) if the election window is open. We don't need anything from the guardians state prior to the election opening, as the keys (submitted by guardians in the setup phase) are just used to construct encrypted ballots to submit to the contract.